File: src/main/java/model/plsol/Plsol.java
package model.plsol;

import java.awt.geom.Point2D;
import java.util.List;

public class Plsol {

    private final Point2D.Double[] points;

    /**
     * @param pu barometric pressure
     * @param tans intake air temperature
     */
    public Plsol(double pu, double tans, double kfurl) {
        points = new Point2D.Double[400];
        double ps = pu;

        for(int i = 0; i < points.length; i++) {
            double x = i;
            double y = plsol(pu, ps, tans, 96, kfurl, x);
            ps = y;

            points[i] = new Point2D.Double(x,y);
        }
    }

    public Plsol(double pu, double tans, double kfurl, List<Double> load) {
        points = new Point2D.Double[load.size()];
        double ps = pu;

        for(int i = 0; i < points.length; i++) {
            double x = load.get(i);
            double y = plsol(pu, ps, tans, 96, kfurl, x);
            ps = y;

            points[i] = new Point2D.Double(x,y);
        }
    }


    public Point2D.Double[] getPoints() {
        return points;
    }

    /**
     *
     * @param pu barometric pressure
     * @param ps intake manifold absolute pressure
     * @param tans intake air temperature
     * @param tmot coolant temperature
     * @param rlsol requested engine load
     * @return
     */
    public static double plsol(double pu, double ps, double tans, double tmot, double kfurl, double rlsol) {
        final double KFPRG = 70.0;
        final double FPBRKDS = 1.016; // VE
        final double KFFWTBR = 0.02;
        final double VPSSPLS = 1.016; // pressure drop over the throttle plate

        double fho = pu/1013.0;

        double pirg = fho*KFPRG;

        double pbr = ps * FPBRKDS;
        double psagr = 250; // ?

        double evtmod = tans + (tmot - tans) * KFFWTBR;
        double fwft = (tans+673.425)/731.334;
        double ftbr = 273.0/(evtmod+273.0)*fwft;

        double fupsrl = kfurl * ftbr; // correct for air temperature (pressure)

        double rfagr = Math.max(pbr-pirg, 0)*fupsrl* psagr/ps; // correct for residual cylinder pressure

        double pssol = (rlsol + rfagr)/fupsrl/FPBRKDS;

        return pssol/VPSSPLS;
    }
}

File: src/main/kotlin/domain/model/plsol/Plsol.kt
package domain.model.plsol

import domain.math.Point

class Plsol {
    val points: Array<Point>

    constructor(pu: Double, tans: Double, kfurl: Double) {
        points = Array(400) { i ->
            Point(0.0, 0.0) // placeholder
        }
        var ps = pu
        for (i in points.indices) {
            val x = i.toDouble()
            val y = plsol(pu, ps, tans, 96.0, kfurl, x)
            ps = y
            points[i] = Point(x, y)
        }
    }

    constructor(pu: Double, tans: Double, kfurl: Double, load: List<Double>) {
        points = Array(load.size) { Point(0.0, 0.0) }
        var ps = pu
        for (i in points.indices) {
            val x = load[i]
            val y = plsol(pu, ps, tans, 96.0, kfurl, x)
            ps = y
            points[i] = Point(x, y)
        }
    }

    companion object {
        fun plsol(pu: Double, ps: Double, tans: Double, tmot: Double, kfurl: Double, rlsol: Double): Double {
            val KFPRG = 70.0
            val FPBRKDS = 1.016
            val KFFWTBR = 0.02
            val VPSSPLS = 1.016

            val fho = pu / 1013.0
            val pirg = fho * KFPRG
            val pbr = ps * FPBRKDS
            val psagr = 250.0

            val evtmod = tans + (tmot - tans) * KFFWTBR
            val fwft = (tans + 673.425) / 731.334
            val ftbr = 273.0 / (evtmod + 273.0) * fwft

            val fupsrl = kfurl * ftbr
            val rfagr = maxOf(pbr - pirg, 0.0) * fupsrl * psagr / ps
            val pssol = (rlsol + rfagr) / fupsrl / FPBRKDS

            return pssol / VPSSPLS
        }
    }
}
